src/core/ddsi/src/q_entity.c

```diff
@@ -4195,6 +4698,46 @@ static void delete_ppt (struct proxy_participant *proxypp, nn_wctime_t timestamp
   gcreq_proxy_participant (proxypp);
 }
 
+#ifdef DDSI_INCLUDE_SECURITY
+
+static void downgrade_to_nonsecure(struct proxy_participant *proxypp)
+{
+  struct proxy_endpoint_common * c;
+  const nn_wctime_t tnow = now();
+
+  ddsrt_mutex_lock (&proxypp->e.lock);
+
+  DDS_CWARNING (&proxypp->e.gv->logconfig, "downgrade participant "PGUIDFMT" to non-secure\n", PGUID (proxypp->e.guid));
+
+  /* Remove security related endpoints. */
+  c = proxypp->endpoints;
+  while (c)
+  {
+    struct proxy_endpoint_common * next = c->next_ep;
+    struct entity_common *e = entity_common_from_proxy_endpoint_common (c);
+    if (is_builtin_security_endpoint(e->guid.entityid))
```

Comment:
If this is the only place ``is_builtin_security_endpoint`` gets used, it would make sense to create a table with entity ids to remove.